python/ql/src/semmle/python/objects/Callables.qll

```diff
@@ -0,0 +1,460 @@
+        exists(Builtin func, BuiltinClassObjectInternal cls |
+            func = this.getBuiltin() and
+            cls = ObjectInternal::fromBuiltin(this.getReturnType()) |
+            obj = TUnknownInstance(cls)
+            or
+            cls = ObjectInternal::noneType() and obj = ObjectInternal::none_()
+            or
+            cls = ObjectInternal::builtin("bool") and obj = ObjectInternal::bool(_)
+        ) and
+        origin = CfgOrigin::unknown()
+    }
+
+    Builtin getReturnType() {
+        exists(Builtin func |
+            func = this.getBuiltin() |
+            ext_rettype(func, result)
+        )
+    }
+
+    override ControlFlowNode getOrigin() {
+        none()
+    }
+
+    override predicate calleeAndOffset(Function scope, int paramOffset) {
+        none()
+    }
+
+    override string getName() {
+        result = this.getBuiltin().getName()
+    }
+
+    override Function getScope() { none() }
+
+    override boolean isDescriptor() { result = true }
+
+    pragma [noinline] override predicate descriptorGetClass(ObjectInternal cls, ObjectInternal value, CfgOrigin origin) {
+        any(ObjectInternal obj).binds(cls, _, this) and
+        value = this and origin = CfgOrigin::unknown()
+    }
+
+    pragma [noinline] override predicate descriptorGetInstance(ObjectInternal instance, ObjectInternal value, CfgOrigin origin) {
+        value = TBoundMethod(instance, this) and origin = CfgOrigin::unknown()
+    }
+
+    override CallNode getACall(PointsToContext ctx) {
+        PointsTo::pointsTo(result.getFunction(), ctx, this, _)
+    }
+
+    override NameNode getParameter(int n) {
+        none()
+    }
+
+    override NameNode getParameterByName(string name) {
+        none()
+    }
+
+    override predicate neverReturns() { none() }
+
+    override predicate functionAndOffset(CallableObjectInternal function, int offset) {
+        function = this and offset = 0
+    }
+
+}
+
+/** Class representing bound-methods.
+ * Note that built-in methods, such as `[].append` are also represented as bound-methods.
+ * Although built-in methods and bound-methods are distinct classes in CPython, their behaviour
+ * is the same and we treat them identically.
+ */
+class BoundMethodObjectInternal extends CallableObjectInternal, TBoundMethod {
+
+    override Builtin getBuiltin() {
+        none()
+    }
+
+    CallableObjectInternal getFunction() {
+        this = TBoundMethod(_, result)
+    }
+
+    ObjectInternal getSelf() {
+        this = TBoundMethod(result, _)
+    }
+
+    override string toString() {
+        result = "Method(" + this.getFunction() + ", " + this.getSelf() + ")"
+    }
+
+    override ObjectInternal getClass() {
+        result = TBuiltinClassObject(Builtin::special("MethodType"))
+    }
+
+    override predicate introducedAt(ControlFlowNode node, PointsToContext context) {
+        none()
+    }
+
+    override boolean testableForEquality() { result = false }
```

Comment:
Because we treat all bound methods of the same object and function as the same, but:
```
>>> class C:
...    def f(self): pass
... 
>>> c = C()
>>> c.f is c.f
False
```